source/g4collimation/G4Interface.cpp

```diff
@@ -154,34 +185,61 @@ extern "C" void g4_set_collimator_(char* name)
 	runManager->ReinitializeGeometry();
 }
 
-extern "C" void g4_collimate_(double* x, double* y, double* xp, double* yp, double* p)
+extern "C" void g4_add_particle_(double* x, double* y, double* xp, double* yp, double* e, int32_t* pdgid, int16_t* nzz, int16_t* naa, int16_t* nqq, double* mass)
 {
 //WARNING: at this stage in SixTrack the units have been converted to GeV, m, and rad!
 //The particle energy input is the TOTAL energy
+//mass (i.e. nucm) is already in MeV!
+
 	double x_in = (*x) * CLHEP::m;
 	double y_in = (*y) * CLHEP::m;
 
 //We want px and py, not the angle!
-	//double px_in = *p * tan(*xp) * CLHEP::GeV;
-	//double py_in = *p * tan(*yp) * CLHEP::GeV;
-	double px_in = (*p) * (*xp) * CLHEP::GeV;
-	double py_in = (*p) * (*yp) * CLHEP::GeV;
-	double p_in = (*p) * CLHEP::GeV;
+	double e_in = (*e) * CLHEP::GeV;
+	double p_in = sqrt((e_in*e_in) - (*mass * *mass));
+
+	double px_in = p_in * (*xp);
+	double py_in = p_in * (*yp);
+
+//	double px_in = p_in * (*xp) * CLHEP::GeV;
+//	double py_in = p_in * (*yp) * CLHEP::GeV;
+
+	G4Stuff in_particle;
+	in_particle.x = x_in;
+	in_particle.y = y_in;
+
+	in_particle.px = px_in;
+	in_particle.py = py_in;
+	in_particle.p = p_in;
 
-//Store the input particle for a check vs the output particle!
-	event->SetInputParticle(x_in, px_in, y_in, py_in, p_in);
+	in_particle.e = e_in;
+	in_particle.pdgid = *pdgid;
+	in_particle.z = *nzz;
+	in_particle.a = *naa;
+	in_particle.q = *nqq;
+	in_particle.id = input_particles.size();
 
+	input_particles.push_back(in_particle);
+}
+
+extern "C" void g4_collimate_()
+{
+	output_particles.clear();
 	//Update the gun with this particle's details
-	part->SetParticleDetails(x_in, y_in, px_in, py_in, p_in);
+	for(size_t n=0; n < input_particles.size(); n++)
+	{
+		part->SetParticleDetails(input_particles.at(n).x, input_particles.at(n).y, input_particles.at(n).px, input_particles.at(n).py, input_particles.at(n).e, input_particles.at(n).p, input_particles.at(n).pdgid, input_particles.at(n).q);
 
-	//Run!
-	runManager->BeamOn(1);
+		//Run!
+		runManager->BeamOn(1);
+	}
+	input_particles.clear();
 }
 
 /**
 * Here we put the particles back into sixtrack and set any flags if needed
 */
-extern "C" void g4_collimate_return_(double* x, double* y, double* xp, double* yp, double* p, int *part_hit, int *part_abs, double *part_impact, double *part_indiv, double *part_linteract)
+extern "C" void g4_collimate_return_(int* j, double* x, double* y, double* xp, double* yp, double* e, int32_t* pdgid, double* m, int16_t* z, int16_t* a, int16_t* q, int *part_hit, int *part_abs, double *part_impact, double *part_indiv, double *part_linteract)
```

Comment:
Why not just pass the whole array pointer?

Reply:
hm, I guess you could but I was worried about possible weirdness going on when doing transfers between fortran and C. I'll rather leave it like this for now, then it is easy to do some per-particle post-processing within the loop.